homeassistant/components/yamaha/config_flow.py

```diff
@@ -0,0 +1,261 @@
+"""Config flow for Yamaha."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
+from urllib.parse import urlparse
+
+from requests.exceptions import ConnectionError
+import rxv
+import voluptuous as vol
+
+from homeassistant import data_entry_flow
+from homeassistant.components import ssdp
+from homeassistant.config_entries import (
+    ConfigEntry,
+    ConfigFlow,
+    ConfigFlowResult,
+    OptionsFlow,
+)
+from homeassistant.const import CONF_HOST, CONF_NAME
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import FlowResultType
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.selector import (
+    SelectOptionDict,
+    Selector,
+    SelectSelector,
+    SelectSelectorConfig,
+    SelectSelectorMode,
+    TextSelector,
+)
+
+from . import get_upnp_serial_and_model
+from .const import (
+    CONF_MODEL,
+    CONF_SERIAL,
+    DEFAULT_NAME,
+    DOMAIN,
+    OPTION_INPUT_SOURCES,
+    OPTION_INPUT_SOURCES_IGNORE,
+)
+from .yamaha_config_info import YamahaConfigInfo
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class YamahaFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a Yamaha config flow."""
+
+    VERSION = 1
+
+    serial_number: str | None = None
+    model: str | None = None
+    host: str
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        # Request user input, unless we are preparing discovery flow
+        if user_input is None:
+            return self._show_setup_form()
+
+        host = user_input[CONF_HOST]
+        serial_number = None
+        model = None
+
+        errors = {}
+        # Check if device is a Yamaha receiver
+        try:
+            info = YamahaConfigInfo(host)
+            await self.hass.async_add_executor_job(rxv.RXV, info.ctrl_url)
+            serial_number, model = await get_upnp_serial_and_model(self.hass, host)
+        except ConnectionError:
+            errors["base"] = "cannot_connect"
+        except Exception:
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+        else:
+            if serial_number is None:
+                errors["base"] = "no_yamaha_device"
```

Comment:
Are there devices without serial number?

Reply:
I am unsure, I did assume there was going to be. I was doubting to add `CONF_HOST` as fallback as the unique_id but that would not accepted I read?
How should I move forward with this?

Reply:
I'm gonna leave this for now, the implementation is the same as the Yamaha Musiccast integration now. I found some SSDP replies through github (https://github.com/foxthefox/yamaha-yxc-nodejs/blob/da9039e3edb1cc338b6f5c19710bbbfb6770b428/lib/data/ssdp_answer_WX30.xml#L17 and https://github.com/foxthefox/yamaha-yxc-nodejs/blob/da9039e3edb1cc338b6f5c19710bbbfb6770b428/lib/data/ssdp_answer_YSP1600.xml#L17).
As well as my RX-A2010

Reply:
@joostlek @slyoldfox yes from my experience there seem to be ssdp responses without serial in the wild, and the rxv lib doesn't support /cant get the serial post ssdp.

Reply:
@pssc does that include crawls from `http://IP/YamahaRemoteControl/desc.xml` ?
@joostlek how should I handle these?

Reply:
@slyoldfox there seems do be a great deal of variability between device implementations I will have a root round tonight and come back but but we could just fall back...
```
For entities that are setup by a config entry, the Config Entry ID can be used as a last resort if no other Unique ID is available.```

Reply:
So config flow yes has specific support for no uid, so this would look to be correct. I think then from https://developers.home-assistant.io/docs/entity_registry_index you can pick the config_flow id for entities or devices

Reply:
@joostlek I have pushed some changes related to devices without serial (mainly devices that don't reply to SSDP or don't reply correctly on the SSDP url).

The legacy behaviour created an instance of `RXV` by only passing the `ctrl_url` value. Which is now also what happens in the fallback scenario when fetching the resource from the SSDP location fails.

The old implementation actually used this behaviour as well if you specified `host: IP` in your configuration.yaml which didn't properly set all the fields (since it only relied on the control url).

So the new implementation favours SSDP detection for the metadata and falls back to a "dumb" control url implementation if SSDP is unavailable.